switchio/apps/measure/cdr.py

```diff
@@ -154,60 +157,77 @@ def on_create(self, sess):
         call_vars['session_count'] = self.pool.count_sessions()
         call_vars['erlangs'] = self.pool.count_calls()
 
-    @event_callback('CHANNEL_ORIGINATE')
-    def on_originate(self, sess):
-        # store local time stamp for originate
-        sess.times['originate'] = sess.time
-        sess.times['req_originate'] = time.time()
-
-    @event_callback('CHANNEL_ANSWER')
-    def on_answer(self, sess):
-        sess.times['answer'] = sess.time
-
-    @event_callback('CHANNEL_HANGUP')
-    def log_stats(self, sess, job):
-        """Append measurement data only once per call
-        """
-        sess.times['hangup'] = sess.time
-        call = sess.call
-
-        if call.sessions:  # still session(s) remaining to be hungup
-            call.caller = call.first
-            call.callee = call.last
-            if job:
-                call.job = job
-            return  # stop now since more sessions are expected to hangup
-
-        # all other sessions have been hungup so store all measurements
-        caller = getattr(call, 'caller', None)
-        if not caller:
-            # most likely only one leg was established and the call failed
-            # (i.e. call.caller was never assigned above)
-            caller = sess
-
-        callertimes = caller.times
-        callee = getattr(call, 'callee', None)
-        calleetimes = callee.times if callee else None
-
-        pool = self.pool
-        job = getattr(call, 'job', None)
-        # NOTE: the entries here correspond to the listed `CDR.fields`
-        rollover = self._ds.append_row((
-            caller.appname,
-            caller['Hangup-Cause'],
-            callertimes['create'],  # invite time index
-            callertimes['answer'],
-            callertimes['req_originate'],  # local time stamp
-            callertimes['originate'],
-            callertimes['hangup'],
-            # 2nd leg may not be successfully established
-            job.launch_time if job else None,
-            calleetimes['create'] if callee else None,
-            calleetimes['answer'] if callee else None,
-            calleetimes['hangup'] if callee else None,
-            pool.count_failed(),
-            call.vars['session_count'],
-            call.vars['erlangs'],
-        ))
-        if rollover:
-            self.log.debug('wrote data to disk')
+        events_sub = list(self.on_create.switchio_events_sub)
```

Comment:
Yes I realize this code is actually longer - so this is an example of where callbacks might be better suited anyway unless you guys can think of a better way to write this.

Reply:
I don't prevent you from maintaining the previous API? Why not support both?

And maybe you can look at supporting `async for` for python 3.6 code:

```python
async for event in await sessioon:
    # ...
```

Reply:
@vodik ok cool. I love the idea of `async for` that's super cool.

Reply:
@vodik yeah only thing I'm thinking right now is that adding `async for` support pushes out `pypy` from the mix. You mind making an issue explaining the API you'd like to see?